heron/healthmgr/src/java/com/twitter/heron/healthmgr/common/TopologyProvider.java

```
@@ -0,0 +1,96 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import com.microsoft.dhalion.events.EventHandler;
+import com.microsoft.dhalion.events.EventManager;
+
+import com.twitter.heron.api.generated.TopologyAPI;
+import com.twitter.heron.api.generated.TopologyAPI.Topology;
+import com.twitter.heron.healthmgr.common.HealthManagerEvents.TopologyUpdate;
+import com.twitter.heron.spi.statemgr.SchedulerStateManagerAdaptor;
+
+import static com.twitter.heron.healthmgr.HealthManager.CONF_TOPOLOGY_NAME;
+
+/**
+ * A topology may be updated after initial deployment. This provider is used to provide the latest
+ * version to any dependent components.
+ */
+@Singleton
+public class TopologyProvider implements Provider<Topology>, EventHandler<TopologyUpdate> {
+  private static final Logger LOG = Logger.getLogger(TopologyProvider.class.getName());
+  private final SchedulerStateManagerAdaptor stateManagerAdaptor;
+  private final String topologyName;
+
+  private Topology topology;
+
+  @Inject
+  public TopologyProvider(SchedulerStateManagerAdaptor stateManagerAdaptor,
+                          EventManager eventManager,
+                          @Named(CONF_TOPOLOGY_NAME) String topologyName) {
+    this.stateManagerAdaptor = stateManagerAdaptor;
+    this.topologyName = topologyName;
+    eventManager.addEventListener(TopologyUpdate.class, this);
+  }
+
+  @Override
+  public synchronized Topology get() {
+    if (topology == null) {
+      fetchLatestTopology();
+    }
+    return topology;
+  }
+
+  private synchronized void fetchLatestTopology() {
+    LOG.log(Level.INFO, "Fetching topology from state manager: {0}", topologyName);
+    this.topology = stateManagerAdaptor.getTopology(topologyName);
+    if (topology == null) {
+      throw new RuntimeException("Failed to fetch topology from the State Manager");
```

Can we throw an exception that is not RuntimeException? Also, let's include the topologyName in the message.